# 1 Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# 2 Import Data

In [2]:
# String for Main Project Folder
path = r'C:\Users\ifari\12-2022 Instacart Basket Analysis'

In [3]:
# String to import Data Folder containing orders.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [4]:
# String to import Data Folder containing products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
# Change datatype of order_id and user_id from orders.csv
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [6]:
# Drop extra index and resetting index
df_ords = df_ords.reset_index(drop = True)
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

# 3 Data Consistency Checks

## 3.0 Mixed Type Data Example

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed datatypes
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

mix  mixed


In [9]:
# Change 'mix' to string datatype
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Recheck for mixed datatypes
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

mix  consistent


## 3.1 Dataframe: df_prods

In [11]:
# Descriptive Statistics about df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

> The max value for 'prices' is 999999 which could be an error or a placeholder for missing values.

In [12]:
# Identify the product_id with the max 'prices' value
df_prods[df_prods.prices == 99999]

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

### 3.1.1 Mixed Data Types

In [13]:
# Check for mixed datatypes in df_prods
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

product_id  consistent
product_name  mixed
aisle_id  consistent
department_id  consistent
prices  consistent


> The 'product_name' containes mixed datatypes because the name of a product could include integers. Otherwise, no other mixed datatypes were found.

### 3.1.2 Missing Values

In [14]:
# Check for missing values in the df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
# Dataframe of ONLY missing values in df_prods
df_prods_nan = df_prods[df_prods['product_name'].isnull() == True]
df_prods_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
# Number of rows and columns of df_prods
df_prods.shape

(49693, 5)

In [17]:
# Dataframe of df_prods without missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [18]:
# Number of rows and columns of df_prods
df_prods_clean.shape

(49677, 5)

### 3.1.2 Duplicates

In [19]:
# Find duplicates in df_prods_clean
df_prods_dups = df_prods_clean[df_prods_clean.duplicated()]
df_prods_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
# Dataframe of df_prods_clean without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

### 3.1.3 Export Dataframes

In [21]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## 3.2 Dataframe: df_ords

In [22]:
# Descriptive Statistics about df_prods
df_ords.describe()

order_number  orders_day_of_week  time_of_order  days_since_prior_order
count  3.421083e+06        3.421083e+06   3.421083e+06            3.214874e+06
mean   1.715486e+01        2.776219e+00   1.345202e+01            1.111484e+01
std    1.773316e+01        2.046829e+00   4.226088e+00            9.206737e+00
min    1.000000e+00        0.000000e+00   0.000000e+00            0.000000e+00
25%    5.000000e+00        1.000000e+00   1.000000e+01            4.000000e+00
50%    1.100000e+01        3.000000e+00   1.300000e+01            7.000000e+00
75%    2.300000e+01        5.000000e+00   1.600000e+01            1.500000e+01
max    1.000000e+02        6.000000e+00   2.300000e+01            3.000000e+01

> The max of 'orders_day_of_week' shouldn't be larger than 7 as there are only 7 days in the week. This should result in a 50th percentile of 3.

### 3.2.1 Mixed Data Types

In [23]:
# Check for mixed datatypes in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col, ' mixed')
  else: print(col, ' consistent')

order_id  consistent
user_id  consistent
order_number  consistent
orders_day_of_week  consistent
time_of_order  consistent
days_since_prior_order  consistent


> No mixed datatypes are found.

### 3.2.2 Missing Values

In [24]:
# Check for missing values in the df_ords
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
time_of_order                  0
days_since_prior_order    206209
dtype: int64

In [25]:
#First 20 rows of df_ords
df_ords.head(20)

order_id user_id  order_number  orders_day_of_week  time_of_order  \
0   2539329       1             1                   2              8   
1   2398795       1             2                   3              7   
2    473747       1             3                   3             12   
3   2254736       1             4                   4              7   
4    431534       1             5                   4             15   
5   3367565       1             6                   2              7   
6    550135       1             7                   1              9   
7   3108588       1             8                   1             14   
8   2295261       1             9                   1             16   
9   2550362       1            10                   4              8   
10  1187899       1            11                   4              8   
11  2168274       2             1                   2             11   
12  1501582       2             2                   5             10   
13  1901567       2             3                   1             10   
14   738281       2             4                   2             10   
15  1673511       2             5                   3             11   
16  1199898       2             6                   2              9   
17  3194192       2             7                   2             12   
18   788338       2             8                   1             15   
19  1718559       2             9                   2              9   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0

In [26]:
# Change datatype of order_id from orders.csv
df_ords['user_id'] = df_ords['user_id'].astype('int64')

# Identify the highest user_id number
df_ords.user_id.max()

206209

In [27]:
# Dataframe of ONLY missing values in df_ors
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]
df_ords_nan

order_id  user_id  order_number  orders_day_of_week  time_of_order  \
0        2539329        1             1                   2              8   
11       2168274        2             1                   2             11   
26       1374495        3             1                   1             14   
39       3343014        4             1                   6             11   
45       2717275        5             1                   3             12   
...          ...      ...           ...                 ...            ...   
3420930   969311   206205             1                   4             12   
3420934  3189322   206206             1                   3             18   
3421002  2166133   206207             1                   6             19   
3421019  2227043   206208             1                   1             15   
3421069  3154581   206209             1                   3             11   

         days_since_prior_order  
0                           NaN  
11                          NaN  
26                          NaN  
39                          NaN  
45                          NaN  
...                         ...  
3420930                     NaN  
3420934                     NaN  
3421002                     NaN  
3421019                     NaN  
3421069                     NaN  

[206209 rows x 6 columns]

> There are a total of 206209 'user_id' numbers which is the same number of null cells contained in df_ords. I believe each null cell corralates to the each 'user_id' first order with Instacart. This would make sense as the first order would not have a value of 'days_since_prior_order'. Due to the first order of each 'user_id' having a null value in 'days_since_prior_order', I would create a 'new_customer' column with true or false values.

### 3.2.3 Duplicates

In [28]:
# Find duplicates in df_ords
df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, time_of_order, days_since_prior_order]
Index: []

> There are no duplicate entries in df_ords

### 3.2.4 Export Dataframe

In [29]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))